### PR TITLE
Arithmetic fast path for addDays/addWeeks in fixed-offset time zones

### PR DESCRIPTION
**Pull request by raimannma**
Adds an arithmetic fast path for `addDays`/`addWeeks` (and thus `subtractDays`/`subtractWeeks`) on `DateTime`/`DateTime64` in fixed-offset time zones (UTC and other zones whose UTC offset never changes within the LUT range). In such zones every calendar day is exactly 86400 seconds, so day/week addition reduces to arithmetic on the unix timestamp and skips the calendar lookup table. Time zones with DST or historical offset changes keep the calendar path unchanged.

Measured ~3x faster for `DateTime` and ~1.35x for `DateTime64` on UTC; no change for DST zones.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Speed up `addDays`, `addWeeks`, `subtractDays`, and `subtractWeeks` on `DateTime` and `DateTime64` values in fixed-offset time zones (such as UTC) by taking an arithmetic fast path.

### CODE REVIEW

File: src/Functions/FunctionDateOrDateTimeAddInterval.h [hunk 1]
```diff
@@ -324,17 +324,23 @@ struct AddDaysImpl
     static NO_SANITIZE_UNDEFINED DateTime64 execute(DateTime64 t, Int64 delta, const DateLUTImpl & time_zone, const DateLUTImpl &, UInt16 scale)
     {
         auto multiplier = DecimalUtils::scaleMultiplier<DateTime64>(scale);
+        if (time_zone.hasFixedOffset())
+            return t + delta * DATE_SECONDS_PER_DAY * multiplier;
```
[the remark this answers is not in the document]
Comment (raimannma) [reply src/Functions/FunctionDateOrDateTimeAddInterval.h]: Good catch, fixed in 54f7a235ae8. The fast path now takes the arithmetic shortcut only when both the input and the result stay inside the LUT (`DateLUTImpl::dayShiftStaysWithinLUT`), and falls back to `DateLUTImpl::addDays` otherwise, for both `addDays` and `addWeeks` and for `DateTime` as well (whose result can also leave the LUT for large deltas). The guard works on the scaled values with overflow-checked arithmetic, so the hot path still performs no division by the scale multiplier.

Boundary behavior is now byte-identical to master — verified against an official 26.7.1.684 build, including the raw values (e.g. `subtractDays(toDateTime64('1900-01-01 00:00:00', 0, 'UTC'), 1)` again clamps to raw `10413705600` = 2300-01-01, and `addDays(…, 9223372036854775807)` keeps the `UInt32` index wraparound), plus a 2M-case randomized sweep asserting fast path == calendar path whenever the guard passes. The stateless test now pins these boundary cases including the raw values that text output would mask, and a gtest covers the new predicate.

The optimization still holds with the guard (same binary, `max_threads=1`, perf-test queries): `addDays` over 100M `DateTime` rows: 0.45s fast path vs 0.61s calendar path (~1.35x); over 50M `DateTime64(3)` rows: 0.98s vs 1.20s (~1.23x) — end-to-end including `numbers()` generation and `sum`.

File: src/Functions/FunctionDateOrDateTimeAddInterval.h [hunk 2]
```diff
@@ -369,17 +375,23 @@ struct AddWeeksImpl
     static NO_SANITIZE_UNDEFINED DateTime64 execute(DateTime64 t, Int64 delta, const DateLUTImpl & time_zone, const DateLUTImpl &, UInt16 scale)
     {
         auto multiplier = DecimalUtils::scaleMultiplier<DateTime64>(scale);
+        if (time_zone.hasFixedOffset() && time_zone.dayShiftStaysWithinLUT(t, delta * 7, multiplier))
```
[the remark this answers is not in the document]
Comment (raimannma) [reply src/Functions/FunctionDateOrDateTimeAddInterval.h]: Fixed in 31e329cc0fa. Added a `weeksToDays` helper that computes the shift once via `__builtin_mul_overflow(delta, 7, &days)`, and all three overloads (`DateTime64`, `Time64`, `UInt32`) now short-circuit the fast path when it reports overflow and reuse the checked `days` value.

On overflow the builtin still writes the wrapped product, so the calendar fallback produces the same result as the previous plain multiplication — only the path selection changes. `04511_add_days_fixed_offset_fast_path` still matches its reference unchanged, including the `addWeeks(..., 9223372036854775807)` case, which now reaches the calendar path and saturates to `9999-12-31` as expected.

[the remark this answers is not in the document]
Comment (raimannma) [reply src/Functions/FunctionDateOrDateTimeAddInterval.h]: Fixed in f02da2fd4bf. The `UInt32` overload now calls `time_zone.addDays(t, days)` on the fallback path, reusing the value `weeksToDays` already computed, so the raw `delta` no longer reaches `DateLUTImpl::addWeeks` and its internal `delta * 7`.

The result is unchanged, since `__builtin_mul_overflow` writes the same wrapped product the plain multiply produced — only the arithmetic is checked, not the outcome.

I also added a `DateTime` huge-delta case to `04511_add_days_fixed_offset_fast_path`, which is why the previous fix looked complete against a green test: the huge-delta `addWeeks` coverage only exercised `DateTime64`, so the overload actually being discussed here had no test at all. The full test now matches its reference with every pre-existing line unchanged, verified against a build rather than assumed.

One thing left deliberately untouched: the `Date` (`UInt16`) and `Date32` (`Int32`) overloads still do a plain `delta * 7`. Those are pre-existing, have no fast path, and are outside this PR's diff, so I have not changed them here.

**Review by yariks5s — APPROVED** [review]

**Comment by raimannma** [comment]
@yariks5s anything left to do for me here?

**Comment by yariks5s** [comment]
> @yariks5s anything left to do for me here?

I think nothing from your side for now. This is a good change, thank you!
